src/cask/scheduler/WorkStealingScheduler.cpp

```diff
@@ -92,44 +92,30 @@ std::vector<std::function<void()>> WorkStealingScheduler::onThreadRequestWork(co
     auto current_thread_id = std::this_thread::get_id();
 
     if (auto self = self_weak.lock()) {
-        // Select a random starting point for work stealing as a simple way to balance work
-        // and avoid lock contention
-        auto random_start_index = std::size_t(std::abs(std::rand())) % self->thread_ids.size();
-        auto i = random_start_index;
-
-        while (true) {
-            // Try and find a scheduler to steal work from
-            auto thread_id = self->thread_ids[i];
-
-            // Skip the current thread (the scheduler thread) to avoid attempting to
-            // steal work from the scheduler requesting the work in the first place
-            if (thread_id != current_thread_id) {
-                // Check if the scheduler exists. If it doesn't then just give up since
-                // we are probably shutting down.
-                auto scheduler_lookup = self->schedulers.find(thread_id);
-                if (scheduler_lookup == self->schedulers.end()) {
-                    break;
-                }
-
-                // Try and steal some work
-                auto batch = scheduler_lookup->second->steal(amount_requested);
-
-                // If a batch was stolen, submit it to the idle scheduler
-                if (batch.size() > 0) {
-                    return batch;
-                }
-            }
-
-            // Move to the next thread - giving up if we've already
-            // checked all threads
-            i = (i + 1) % self->thread_ids.size();
-            if (i == random_start_index) {
-                break;
-            }
+        // Select a random starting scheduler to attempt to steal work from
+        auto scheduler_index = std::size_t(std::abs(std::rand())) % self->thread_ids.size();
+        auto thread_id = self->thread_ids[scheduler_index];
+
+        // If we selected the current scheduler, go ahead and skip it
+        if (thread_id == current_thread_id) {
+            scheduler_index = (scheduler_index + 1) % self->thread_ids.size();
+            thread_id = self->thread_ids[scheduler_index];
+        }
```

Comment:
This means the thread after this one has 2x the chance of getting chosen vs any other thread. Probably fine. Just pointing it out.

Reply:
Ya, that's right. Good catch. I can adjust the maths to prevent this.

Reply:
I added a small map to also allow fast reverse lookup of the scheduler's index from the current thread ID. That allows this math to be trivially changed to avoid this problem by rolling for n-1 schedulers and just adjusting the index as-needed to skip the scheduler that's currently trying to steal.